src/System.Management.Automation/engine/lang/parserutils.cs

```diff
@@ -978,15 +980,15 @@ internal static object ReplaceOperator(ExecutionContext context, IScriptExtent e
                     lvalString = lval?.ToString() ?? string.Empty;
                 }
 
-                return ReplaceOperatorImpl(context, lvalString, rr, substitute);
+                return ReplaceOperatorImpl(context, lvalString, rr, substitute, ref cachedReplacementString, ref cachedMatchEvaluator);
             }
             else
             {
                 List<object> resultList = new List<object>();
                 while (ParserOps.MoveNext(context, errorPosition, list))
                 {
                     string lvalString = PSObject.ToStringParser(context, ParserOps.Current(errorPosition, list));
-                    resultList.Add(ReplaceOperatorImpl(context, lvalString, rr, substitute));
+                    resultList.Add(ReplaceOperatorImpl(context, lvalString, rr, substitute, ref cachedReplacementString, ref cachedMatchEvaluator));
```

Comment:
So the issues I have here are:

- In this top call, we'll never use the cached value. So we've declared null variables too high up just to pass them into a method that won't use them.
- Passing `ref` parameters into a method that returns a value puts cognitive load on the reader that isn't really necessary here.
- The cached values are essentially internal to the method implementation that we're calling, and we're forcing the caller to own them.

Rather than us building out increasingly complex static methods that leak their implementation with more and more parameters, we could use an object-oriented approach:

```csharp
    private struct OperatorReplacer
    {
        public static OperatorReplacer Create(ExecutionContext context)
        {
            return new OperatorReplacer(context, replacementString: null, matchEvaluator: null);
        }

        private readonly ExecutionContext _context;

        private string _cachedReplacementString;

        private MatchEvaluator _cachedMatchEvaluator;

        private OperatorReplacer(
            ExecutionContext context,
            string replacementString,
            MatchEvaluator matchEvaluator)
        {
            _context = context;
            _cachedReplacementString = replacementString;
            _cachedMatchEvaluator = matchEvaluator;
        }

        public object ReplaceOperator(string input, Regex regex, object substitute)
        {
            if (_cachedReplacementString is not null)
            {
                return regex.Replace(input, _cachedReplacementString);
            }

            if (_cachedMatchEvaluator is not null)
            {
                return regex.Replace(input, _cachedMatchEvaluator);
            }

            ...
        }
    }
```

That way all the caching and the other implementation is contained within the struct (I've written a struct since I'm assuming we won't want to allocate, although the object is less cumbersome as a class).

The other advantage is that we clean up the `ReplaceOperatorImpl` call a bit, so it only refers to the things that are relevant to each call.

Within the parent `ReplaceOperator` method then, all we have to do is:

```csharp
var replacer = OperatorReplacer.Create(context);
...
return replacer.ReplaceOperator(lvalString, rr, substitute);
```

Reply:
It is performance PR - this assumes minimal refactoring. 
You proposal looks interesting but I'd postpone it for follow PR from you or me.

Reply:
Given how many PRs we're juggling right now, I'd prefer to not go into followup PRs. Each PR needs to monotonically improve the code.

Reply:
Since you already implemented the code I added it :-)

Reply:
Well now I understand the substitute logic better, I've written another approach that means we don't need any of the caching concepts at all

Reply:
I understand that you want a performance PR to not make serious changes, but to my mind the code that pre-existed here was, simply put, bad.

Trying to improve the performance of bad code by adding things to it just puts us deeper in the hole. And PowerShell already has so much hard-to-maintain code.

I think if we're going to fix the problem (and we've already invested the time now), we may as well make the change we ultimately want to make and ensure it's the right one with testing.